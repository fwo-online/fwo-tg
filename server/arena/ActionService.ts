import arena from '@/arena';
import type { Action } from '@fwo/shared';

export type ActionKey = keyof typeof arena.actions;

export class ActionService {
  static isAction(maybeAction: string): maybeAction is ActionKey {
    return maybeAction in arena.actions;
  }

  static isBaseAction(action: string) {
    if (ActionService.isAction(action)) {
      return arena.actions[action].lvl === 0;
    }

    return false;
  }

  static toObject(actionKey: ActionKey, lvl = 1): Action {
    const action = arena.actions[actionKey];

    const actionObject: Action = {
      name: action.name,
      displayName: action.displayName,
      orderType: action.orderType,
    };

    if ('cost' in action) {
      actionObject.cost = Array.isArray(action.cost) ? action.cost[lvl - 1] : action.cost;
    }

    if ('costType' in action) {
      actionObject.costType = action.costType;
    }

    if ('proc' in action) {
      actionObject.power = action.proc;
    }

    return actionObject;
  }
}
