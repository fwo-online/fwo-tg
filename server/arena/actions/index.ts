export { default as attack } from './attack';
export { default as handsHeal } from './handsHeal';
export { default as protect } from './protect';
export { default as regeneration } from './regeneration';
