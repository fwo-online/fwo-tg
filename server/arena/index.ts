import type { Clan } from '@/models/clan';
import type { ItemOutput } from '@fwo/shared';
import type { ItemSet } from '@fwo/shared';
import type { CharacterService } from './CharacterService';
import type GameService from './GameService';
import type MatchMaking from './MatchMakingService';
import type * as actions from './actions';
import type * as magics from './magics';
import type * as passiveSkills from './passiveSkills';
import type * as skills from './skills';
import type * as weaponMastery from './weaponMastery';

export default {
  mm: null as unknown as typeof MatchMaking,
  characters: {} as Record<string, CharacterService>,
  games: {} as Partial<Record<string, GameService>>,
  items: {} as unknown as Record<string, ItemOutput>,
  itemsSets: {} as unknown as Record<string, ItemSet>,
  magics: {} as unknown as typeof magics,
  skills: {} as unknown as typeof skills,
  actions: {} as unknown as typeof actions &
    typeof magics &
    typeof skills &
    typeof passiveSkills &
    typeof weaponMastery,
  clans: new Map<string, Clan>(),
};
