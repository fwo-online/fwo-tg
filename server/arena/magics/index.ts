/**
 * Первичный загрузчик магий
 */
export { default as acidSpittle } from './acidSpittle';
export { default as anathema } from './anathema';
export { default as blessing } from './blessing';
export { default as blight } from './blight';
export { default as bodySpirit } from './bodySpirit';
export { default as chainLightning } from './chainLightning';
export { default as curse } from './curse';
export { default as dispel } from './dispel';
export { default as dustShield } from './dustShield';
export { default as eclipse } from './eclipse';
export { default as entangle } from './entangle';
export { default as exorcism } from './exorcism';
export { default as fireBall } from './fireBall';
export { default as frostTouch } from './frostTouch';
export { default as fireRain } from './fireRain';
export { default as glitch } from './glitch';
export { lightHeal } from './lightHeal';
export { default as lightShield } from './lightShield';
export { default as madness } from './madness';
export { default as magicArmor } from './magicArmor';
export { default as magicArrow } from './magicArrow';
export { default as magicDefense } from './magicDefense';
export { default as mediumAura } from './mediumAura';
export { default as mediumHeal } from './mediumHeal';
export { default as magicWall } from './magicWall';
export { default as paralysis } from './paralysis';
export { default as poisonBreath } from './poisonBreath';
export { default as physicalSadness } from './physicalSadness';
export { default as rockfall } from './rockfall';
export { default as silence } from './silence';
export { default as smallAura } from './smallAura';
export { default as stoneSkin } from './stoneSkin';
export { default as strongHeal } from './strongHeal';
export { default as secondLife } from './secondLife';
export { default as sleep } from './sleep';
export { default as strongAura } from './strongAura';
export { default as vampirism } from './vampirism';
