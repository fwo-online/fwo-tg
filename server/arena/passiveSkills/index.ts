export { default as divineWill } from './divineWill';
export { default as fatesMiss } from './fatesMiss';
export { default as lacerate } from './lacerate';
export { default as markOfDarkness } from './markOfDarkness';
export { default as nightcall } from './nightcall';
export { default as staticProtect } from './staticProtect';
