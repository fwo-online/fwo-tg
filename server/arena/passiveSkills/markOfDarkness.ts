import type { Affect, AffectFn } from '@/arena/Constuructors/interfaces/Affect';
import { PassiveSkillConstructor } from '@/arena/Constuructors/PassiveSkillConstructor';

export class MarkOfDarkness extends PassiveSkillConstructor implements Affect {
  constructor() {
    super({
      name: 'markOfDarkness',
      displayName: '🌑 Метка тьмы',
      description: 'Накладывает метку на врагов во время затмения',
      chance: [80, 85, 90, 95, 100],
      effect: [1.2, 1.4, 1.6, 1.8, 2],
      bonusCost: [],
    });
  }

  run() {
    this.initiator.proc *= this.getEffect();
  }

  preAffect?: AffectFn | undefined = (context) => {
    this.applyContext(context);

    if (!this.isActive()) {
      return;
    }

    if (!this.target.flags.isMarkedByDarkness.length) {
      return;
    }

    this.run();

    return this.target.flags.isMarkedByDarkness.map(({ initiator }) => {
      return this.getSuccessResult({ ...this.params, initiator });
    });
  };

  postAffect?: AffectFn | undefined = (context) => {
    this.applyContext(context);
    this.swapParams();

    if (!this.isActive()) {
      return;
    }

    if (!this.checkChance()) {
      return;
    }

    if (!this.game.flags.global.isEclipsed.length) {
      return;
    }

    return this.game.players.getAliveEnemies(this.initiator).map((target) => {
      target.flags.isMarkedByDarkness.push({ initiator: this.initiator, val: 0 });
      return this.getSuccessResult({
        ...this.params,
        target,
      });
    });
  };
}

export default new MarkOfDarkness();
