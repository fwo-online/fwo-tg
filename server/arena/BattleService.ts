import { Markup } from 'telegraf';
import BattleKeyboard from '@/helpers/BattleKeyboard';
import * as channelHelper from '../helpers/channelHelper';
import OrderError from './errors/OrderError';
import type Game from './GameService';
import arena from './index';

type BattleReply = {
  message: string;
  keyboard: ReturnType<typeof Markup.button.callback>[][];
}

/**
 * –ó–∞–∫–∞–∑—ã–≤–∞–µ–º –≤ —Ç—Ä–∏ —ç—Ç–∞–ø–∞:
 * 1. –í—ã–±–æ—Ä —É–º–µ–Ω–∏—è (action_{attack})
 * 2. –í—ã–±–æ—Ä —Ü–µ–ª–∏ ({attack}_{target})
 * 3. –í—ã–±–æ—Ä —Å–∏–ª—ã ({attack}_{target}_{proc})
 */

const getActions = () => ({ ...arena.actions, ...arena.skills, ...arena.magics });

const getText = {
  order: (proc: number) => `–£ —Ç–µ–±—è –æ—Å—Ç–∞–ª–æ—Å—å *${proc}%*
–ó–∞–∫–∞–∑—ã:\n`,
  proc: (displayName: string, target: string) => `–í—ã–±–µ—Ä–∏ —Å–∏–ª—É _${displayName}_ –Ω–∞ ${target ? `–∏–≥—Ä–æ–∫–∞ ${target}` : '—Å–µ–±—è'}`,
  target: (displayName: string) => `–í—ã–±–µ—Ä–∏ —Ü–µ–ª—å –¥–ª—è _${displayName}_`,
};

function getCurrentOrders(charId: string, game: Game) {
  return game.orders.getPlayerOrders(charId)
    .map((o) => {
      if (o.target === charId) {
        return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) –Ω–∞ *—Å–µ–±—è*`;
      }
      const target = game.players.getById(o.target);
      if (!target) {
        throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
      }
      return `_${getActions()[o.action].displayName}_ (*${o.proc}%*) –Ω–∞ –∏–≥—Ä–æ–∫–∞ *${target.nick}*`;
    })
    .join('\n');
}

function getTargetKeyboard(charId: string, game: Game, action: string) {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const orders = game.orders.getPlayerOrders(charId);
  const { orderType } = getActions()[action];
  const proc = arena.skills[action] ? `_${arena.skills[action].proc}` : '';

  return game.players.alivePlayers
    .filter((target) => {
      if (orderType === 'enemy') {
        return !player.isAlly(target);
      }
      if (orderType === 'team') {
        return player.isAlly(target);
      }
      if (orderType === 'teamExceptSelf') {
        return player.isAlly(target, false);
      }

      return !orders.some((order) => target.id === order.target && action === order.action);
    })
    .map(({ nick, id }) => Markup.button.callback(nick, `${action}_${id}${proc}`));
}

/**
 * –í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –∫–Ω–æ–ø–∫–∏ —Å –ø—Ä–æ—Ü–µ–Ω—Ç–æ–º –∑–∞–∫–∞–∑–∞
 */
function getPercentKeyboard(action: string, target: string, proc: number) {
  return Array
    .from(new Set([5, 10, 25, 50, 75, proc]))
    .filter((key) => key <= proc)
    .map((key) => Markup.button.callback(key.toString(), `${action}_${target}_${key}`));
}

/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è –ø–µ—Ä–≤–æ–≥–æ —ç—Ç–∞–ø–∞
 * @param charId
 * @param game
 */
function orderMessage(charId: string, game: Game): BattleReply {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const message = getText.order(player.proc)
    .concat(getCurrentOrders(charId, game));
  const keyboard = channelHelper.getOrderButtons(player);
  return { message, keyboard };
}

/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è –≤—Ç–æ—Ä–æ–≥–æ —ç—Ç–∞–ø–∞
 */
function targetMessage(charId: string, game: Game, action: string): BattleReply {
  const { displayName } = getActions()[action];
  const message = getText.target(displayName);
  const keyboard = [
    getTargetKeyboard(charId, game, action),
    [Markup.button.callback('üîô –ù–∞–∑–∞–¥', 'back')],
  ];
  return { message, keyboard };
}

type PercentMessageParams = {
  initiator: string;
  game: Game;
  action: string;
  target: string;
  isSelfAction?: boolean;
}
/**
 * –°–æ–æ–±—â–µ–Ω–∏–µ –¥–ª—è —Ç—Ä–µ—Ç—å–µ–≥–æ —ç—Ç–∞–ø–∞
 */
function percentMessage({
  initiator: initiatorId, game, action, target: targetId, isSelfAction = false,
}: PercentMessageParams) {
  const initiator = game.players.getById(initiatorId);
  if (!initiator) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const target = game.players.getById(targetId);
  if (!target) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const { displayName } = getActions()[action];
  const backButtonData = isSelfAction ? 'back' : `action_${action}`;
  const message = getText.proc(displayName, isSelfAction ? '' : target.nick);
  const keyboard = [
    getPercentKeyboard(action, targetId, initiator.proc),
    [Markup.button.callback('üîô –ù–∞–∑–∞–¥', backButtonData)],
  ];
  return { message, keyboard };
}

export function getDefaultMessage(charId: string, game: Game): BattleReply {
  return orderMessage(charId, game);
}

export function getAllMagicsMessage(charId: string, game: Game): BattleReply {
  const player = game.players.getById(charId);
  if (!player) {
    throw new OrderError('–ò–≥—Ä–æ–∫ –Ω–µ –Ω–∞–π–¥–µ–Ω');
  }
  const message = '–°–ø–∏—Å–æ–∫ –≤—Å–µ—Ö –º–∞–≥–∏–π';

  const keyboard = [
    ...new BattleKeyboard(player).setMagics(true).render(true),
    [Markup.button.callback('–ù–∞–∑–∞–¥', 'back')],
  ];

  return { message, keyboard };
}
/**
   * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω–æ–≥–æ –¥–µ–π—Å—Ç–≤–∏—è (–ø–µ—Ä–≤—ã–π —ç—Ç–∞–ø)
   */
export function handleAction(initiator: string, game: Game, action: string): BattleReply {
  if (action === 'repeat') return repeatOrder(initiator, game);
  if (action === 'reset') return resetOrder(initiator, game);

  const { orderType } = getActions()[action];
  const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

  // –ï—Å–ª–∏ —É–º–µ–Ω–∏–µ –Ω–∞–ø—Ä–∞–≤–ª–µ–Ω–æ –Ω–∞ —Å–µ–±—è –∏ –∏–º–µ–µ—Ç –ø—Ä–æ—Ü–µ–Ω—Ç (—É–º–µ–Ω–∏—è), —Ç–æ –¥–µ–ª–∞–µ–º –∑–∞–∫–∞–∑ —Å—Ä–∞–∑—É
  if (orderType === 'self') {
    if (typeof proc !== 'undefined') {
      game.orders.orderAction({
        initiator, target: initiator, action, proc,
      });
      return orderMessage(initiator, game);
    }
    return percentMessage({
      initiator, game, action, target: initiator, isSelfAction: true,
    });
  }
  return targetMessage(initiator, game, action);
}
/**
 * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω–æ–π —Ü–µ–ª–∏ (–≤—Ç–æ—Ä–æ–π —ç—Ç–∞–ø)
 */
// eslint-disable-next-line max-len
export function handleTarget(initiator: string, game: Game, action: string, target: string): BattleReply {
  const proc = arena.skills[action] ? arena.skills[action].proc : undefined;

  if (typeof proc !== 'undefined') {
    game.orders.orderAction({
      initiator, target, action, proc,
    });
    return orderMessage(initiator, game);
  }
  return percentMessage({
    initiator, game, action, target,
  });
}
/**
 * –û–±—Ä–∞–±–æ—Ç–∫–∞ –≤—ã–±—Ä–∞–Ω–Ω—ã—Ö –ø—Ä–æ—Ü–µ–Ω—Ç–æ–≤ (—Ç—Ä–µ—Ç–∏–π —ç—Ç–∞–ø)
 */
// eslint-disable-next-line max-len
export function handlePercent(charId: string, game: Game, action: string, target: string, proc: number): BattleReply {
  game.orders.orderAction({
    initiator: charId, target, action, proc,
  });
  return orderMessage(charId, game);
}

export function repeatOrder(charId: string, game: Game): BattleReply {
  game.orders.repeatLastOrder(charId);
  return orderMessage(charId, game);
}

export function resetOrder(charId: string, game: Game): BattleReply {
  game.orders.resetOrdersForPlayer(charId);
  return orderMessage(charId, game);
}
