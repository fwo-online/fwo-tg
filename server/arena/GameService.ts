import { createGame } from '@/api/game';
import type { Game } from '@/models/game';
import type { LongItem } from '@/arena/Constuructors/LongMagicConstructor';
import { engine } from '@/arena/EngineService';
import { HistoryService, type HistoryItem } from '@/arena/HistoryService';
import type * as magics from '@/arena/magics';
import OrderService from '@/arena/OrderService';
import PlayersService, { type Player } from '@/arena/PlayersService';
import { RoundService, RoundStatus } from '@/arena/RoundService';
import arena from '@/arena';
import { mapValues } from 'es-toolkit';
import EventEmitter from 'node:events';
import { type GameStatus, reservedClanName } from '@fwo/shared';
import { getRandomComponent } from '@/utils/getRandomComponent';

export type KickReason = 'afk' | 'run';

export interface GlobalFlags {
  isEclipsed: { initiator: Player }[];
}

/**
 * GameService
 *
 * @description Обработка около игровой логики
 * @module Service/Game
 * @todo сейчас после того как Player отключился, socket выходит из room.
 * Нужен механизм подключения обратно, если клиент "обновил" страницу или
 * переподключился к игре после disconnect(разрыв соединения)
 */

/**
 * Класс для объекта игры
 */
export default class GameService extends EventEmitter<{
  end: [
    {
      reason: string | undefined;
      statistic: Partial<Record<string, { exp: number; gold: number; nick: string }[]>>;
    },
  ];
  startOrders: [];
  endOrders: [];
  preKick: [{ reason: string; player: Player }];
  kick: [{ reason: string; player: Player }];
  startRound: [{ round: number; status: Record<string, GameStatus[]> }];
  endRound: [{ dead: Player[]; log: HistoryItem[] }];
}> {
  players: PlayersService;
  orders: OrderService;
  round = new RoundService();
  history = new HistoryService();
  longActions: Partial<Record<keyof typeof magics, LongItem[]>> = {};
  info!: Game;
  flags: {
    noDamageRound: number;
    global: GlobalFlags;
  };

  /**
   * Конструктор объекта игры
   * @param players массив игроков
   */
  constructor(players: string[]) {
    super();

    this.players = new PlayersService(players);
    this.orders = new OrderService(this.players, this.round);
    this.flags = {
      noDamageRound: 0,
      global: {
        isEclipsed: [],
      },
    };
  }

  /**
   * Функция проверки окончания игры
   */
  get isGameEnd(): boolean {
    return (
      this.isTeamWin ||
      this.players.alivePlayers.length === 0 ||
      this.flags.noDamageRound > 2 ||
      this.round.count > 9
    );
  }

  get isTeamWin(): boolean {
    const { [reservedClanName]: noClan, ...groupByClan } = this.players.groupByClan(
      this.players.alivePlayers,
    );
    if (!noClan?.length) {
      return Object.keys(groupByClan).length === 1;
    }
    return noClan.length === 1 && !Object.keys(groupByClan).length;
  }

  getEndGameReason() {
    if (this.flags.noDamageRound > 2) {
      return 'noDamageRound';
    }
  }

  get checkRoundDamage(): boolean {
    return !!this.history.hasDamageForRound(this.round.count);
  }

  /**
   * Предзагрузка игры
   */
  preLoading(): void {
    this.initHandlers();
    this.startGame();

    arena.games[this.info.id] = this;

    this.info.players.forEach((playerId) => {
      arena.characters[playerId].gameId = this.info.id;
    });
    // @todo add statistic +1 game for all players
  }

  /**
   * Старт игры
   */
  startGame(): void {
    console.debug('GC debug:: startGame', 'gameId:', this.info.id);
    this.round.initRound();
  }

  /**
   *@todo Остановка игры
   */
  pauseGame(): void {
    console.debug(this.info.id);
  }

  /**
   * Прекик, помечаем что пользователь не выполнил заказ и дальше будет выброшен
   * @param id id игрока, который будет помочен как бездействующий
   * @param reason строка, подставляющаяся в флаг isKicked
   */
  preKick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: preKick', id, 'no player');
      return;
    }
    player.preKick(reason);
    this.emit('preKick', { reason, player });
  }

  /**
   * Функция "выброса игрока" из игры без сохранения накопленных статов
   * @param id id игрока, который будет выброшен
   * @param reason причина кика
   */
  kick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: kick', id, 'no player');
      return;
    }

    this.emit('kick', { reason, player });

    const char = arena.characters[id];
    char.addGameStat({ runs: 1 });
    void char.saveToDb();
    char.autoreg = false;
    this.players.kick(id);
    this.info.players.splice(this.info.players.indexOf(id), 1);
    this.cleanLongMagics();
    char.gameId = '';
  }

  checkRun(player: Player) {
    if (player.flags.isKicked === 'run') {
      this.kick(player.id, player.flags.isKicked);
      return;
    }
  }

  /**
   * Проверяем делал ли игрок заказ. Помечает isKicked, если нет
   * @param player
   */
  checkOrders(player: Player): void {
    if (!player.alive || this.round.count === 1) {
      return;
    }

    if (!this.orders.checkPlayerOrderLastRound(player.id)) {
      if (player.flags.isKicked === 'afk') {
        this.kick(player.id, 'afk');
      } else {
        this.preKick(player.id, 'afk');
      }
    } else {
      /** @todo create clear kick flag method */
      player.preKick(undefined);
    }
  }

  /**
   * Ставим флаги, влияющие на окончание игры
   */
  handleEndGameFlags(): void {
    if (this.checkRoundDamage) {
      this.flags.noDamageRound = 0;
    } else {
      this.flags.noDamageRound += 1;
    }
  }

  recordOrderResult(item: HistoryItem) {
    this.history.addHistoryForRound(item, this.round.count);
  }

  getLastRoundResults() {
    return this.history.getHistoryForRound(this.round.count - 1);
  }

  getRoundResults() {
    return this.history.getHistoryForRound(this.round.count);
  }

  /**
   * @description Завершение игры
   *
   */
  endGame(): void {
    console.log('GC debug:: endGame', this.info.id);
    // Отправляем статистику
    setTimeout(async () => {
      const statistic = this.statistic();
      await this.saveGame();
      this.forAllPlayers((player: Player) => {
        arena.characters[player.id].gameId = '';
      });

      this.emit('end', { reason: this.getEndGameReason(), statistic });
      this.removeAllListeners();

      arena.mm.cancel();

      // FIXME move to client side
      this.forAllPlayers((player: Player) => {
        const char = arena.characters[player.id];
        if (char.expEarnedToday >= char.expLimitToday) {
          char.autoreg = false;
        }
        if (!char.autoreg) return;
        arena.mm.push({
          id: player.id,
          psr: 1000,
          startTime: Date.now(),
        });
      });
    }, 5000);
  }

  /**
   * Создание объекта в базе // потребуется для ведения истории
   * @return Объект созданный в базе
   */
  async createGame() {
    try {
      const dbGame = await createGame(this.players.init);
      this.info = dbGame;
      this.info.id = this.info._id.toString();
      return this;
    } catch (e) {
      console.log('GC debug:: createGame', e);
    } finally {
      this.preLoading();
    }
  }

  /**
   * Очищаем глобальные флаги в бою
   * затмение, бунт богов, и т.п
   */
  refreshRoundFlags(): void {
    this.flags.global.isEclipsed = [];
  }

  /**
   * Подвес
   */
  initHandlers(): void {
    // Обработка сообщений от Round Module
    this.round.subscribe(({ state, round }) => {
      switch (state) {
        case RoundStatus.START_ROUND: {
          this.forAllPlayers(this.checkOrders);
          this.sendStatus(round);
          break;
        }
        case RoundStatus.END_ROUND: {
          this.forAllPlayers(this.checkRun);
          this.emit('endRound', { dead: this.sortDead(), log: this.getRoundResults() });
          this.players.reset();
          this.orders.reset();
          this.handleEndGameFlags();
          if (this.isGameEnd) {
            this.round.unsubscribe();
            this.endGame();
          } else {
            this.refreshRoundFlags();
            this.round.nextRound();
          }
          break;
        }
        case RoundStatus.ENGINE: {
          engine(this);
          break;
        }
        case RoundStatus.START_ORDERS: {
          this.emit('startOrders');
          break;
        }
        case RoundStatus.END_ORDERS: {
          this.emit('endOrders');
          break;
        }
        default: {
          console.log('InitHandler:', state, 'undef event');
        }
      }
    });
  }

  /**
   * Метод сохраняющий накопленную статистику игроков в базу и сharObj
   * @todo нужен общий метод сохраняющий всю статистику
   */
  async saveGame() {
    try {
      const promises = this.info.players.map(async (id) => {
        const player = this.players.getById(id);
        if (!player) {
          return;
        }
        const isWinner = player.alive;

        await arena.characters[id].resources.addResources({
          gold: player.stats.collect.gold,
          exp: player.stats.collect.exp,
          components: player.stats.collect.component
            ? { [player.stats.collect.component]: 1 }
            : undefined,
        });

        const kills = this.players.getKills(id).length;
        const death = isWinner ? 0 : 1;

        arena.characters[id].addGameStat({
          games: 1,
          death,
          kills,
        });
        return arena.characters[id].saveToDb();
      });

      await Promise.all(promises);
    } catch (e) {
      console.log('Game:', e);
    }
  }

  /**
   * Функция послематчевой статистики
   * @todo надо как-то переделать функцию, потому что непонятно, что она мутирует данные
   * @return возвращает строку статистики по всем игрокам
   */
  statistic() {
    this.giveGoldforKill();
    const winners = this.players.alivePlayers;
    const gold = this.players.deadPlayers.length ? 5 : 1;
    winners.forEach((p) => {
      p.stats.addGold(gold);
      p.stats.addComponent(getRandomComponent(60));
    });

    this.players.deadPlayers.forEach((p) => {
      p.stats.addComponent(getRandomComponent(20));
    });

    const playersByClan = this.players.groupByClan();

    const getStatisticString = (p: Player) => ({
      nick: p.nick,
      exp: p.stats.collect.exp,
      gold: p.stats.collect.gold,
      component: p.stats.collect.component,
    });

    const statisticByClan = mapValues(playersByClan, (players) => {
      return players?.map(getStatisticString);
    });

    return statisticByClan;
  }

  /**
   * Функция пробегает всех убитых и раздает золото убийцам
   */
  giveGoldforKill(): void {
    this.players.deadPlayers.forEach((p) => {
      const killer = this.players.getById(p.getKiller());
      if (killer && killer.id !== p.id) {
        killer.stats.addGold(5 * p.lvl);
      }
    });
  }

  /**
   * Функция выставляет "смерть" для игроков имеющих hp < 0;
   * Отсылает сообщение о смерти игрока в последнем раунде
   * @todo сообщение о смерти как-то нормально нужно сделать,
   * чтобы выводило от чего и от кого умер игрок
   */
  sortDead() {
    const dead = this.players.sortDead();
    this.cleanLongMagics();
    return dead;
  }

  /**
   * Очистка массива длительных магий от умерших
   */
  cleanLongMagics(): void {
    this.longActions = mapValues(this.longActions, (longMagicType) => {
      return longMagicType?.filter((act) => {
        return this.players.getById(act.target)?.alive;
      });
    });
  }

  /**
   * Интерфейс для работы со всеми игроками в игре
   * @param f функция применяющая ко всем игрокам в игре
   */
  forAllPlayers(f: (player: Player) => void): void {
    this.players.players.forEach((p) => f.call(this, p));
  }

  getStatus() {
    const playersByClan = this.players.groupByClan();

    return mapValues(playersByClan, (players = []) => {
      return players.map((p) => p.getStatus());
    });
  }

  /**
   * Рассылка состояний живым игрокам
   */
  sendStatus(round: number): void {
    const status = this.getStatus();

    this.emit('startRound', { round, status });
  }
}
