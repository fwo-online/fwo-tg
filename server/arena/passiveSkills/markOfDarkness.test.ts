import { afterEach, beforeAll, beforeEach, describe, expect, it, spyOn } from 'bun:test';
import casual from 'casual';
import GameService from '@/arena/GameService';
import { eclipse, vampirism } from '@/arena/magics';
import type { Char } from '@/models/character';
import TestUtils from '@/utils/testUtils';
import markOfDarkness from './markOfDarkness';

// npm t server/arena/passiveSkills/markOfDarkness.test.ts

describe('markOfDarkness', () => {
  let game: GameService;
  let initiator: Char;
  let target: Char;

  beforeAll(async () => {
    casual.seed(1);
    eclipse.registerPostAffects([markOfDarkness]);

    initiator = await TestUtils.createCharacter({
      passiveSkills: { markOfDarkness: 3 },
      magics: { vampirism: 3 },
    });
    target = await TestUtils.createCharacter({
      magics: { eclipse: 3 },
    });
  });

  beforeEach(async () => {
    game = new GameService([initiator.id, target.id]);
    spyOn(global.Math, 'random').mockReturnValue(0.5);
  });

  afterEach(() => {
    spyOn(global.Math, 'random').mockRestore();
  });

  it('should apply mark', () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('mp', 100);
    game.players.players[1].proc = 1;
    game.players.players[1].stats.set('mp', 100);

    eclipse.cast(game.players.players[1], game.players.players[0], game);
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
    expect(game.players.players[1].flags.isMarkedByDarkness).toHaveLength(1);
  });

  it('mark should increase effect', () => {
    game.players.players[0].proc = 1;
    game.players.players[0].stats.set('magic.attack', 100);
    game.players.players[0].stats.set('mp', 100);
    game.players.players[1].flags.isMarkedByDarkness.push({
      initiator: game.players.players[0],
      val: 0,
    });

    vampirism.cast(game.players.players[0], game.players.players[1], game);
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();

    vampirism.registerPreAffects([markOfDarkness]);
    vampirism.cast(game.players.players[0], game.players.players[1], game);
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
