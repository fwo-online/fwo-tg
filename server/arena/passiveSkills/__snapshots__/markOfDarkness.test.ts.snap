// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`markOfDarkness should apply mark 1`] = `
"*corporis* –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–ª _–ó–∞—Ç–º–µ–Ω–∏–µ_ –Ω–∞ *quisquam*
\\[ üìñ80 ]
_üåë –ú–µ—Ç–∫–∞ —Ç—å–º—ã_ *quisquam*"
`;

exports[`markOfDarkness mark should increase effect 1`] = `
"*quisquam* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *corporis* –Ω–∞–Ω–µ—Å—è *6.3* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *6.3* –∑–¥–æ—Ä–æ–≤—å—è
\\[ corporis ‚ú® üíî-6.3/1.7 üìñ60 ]"
`;

exports[`markOfDarkness mark should increase effect 2`] = `
"*quisquam* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *corporis* –Ω–∞–Ω–µ—Å—è *6.3* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *6.3* –∑–¥–æ—Ä–æ–≤—å—è
\\[ corporis ‚ú® üíî-6.3/1.7 üìñ60 ]
*quisquam* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *corporis* –Ω–∞–Ω–µ—Å—è *10.08* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *1.7* –∑–¥–æ—Ä–æ–≤—å—è
\\[ corporis ‚ú® üíî-10.08/-8.38 üìñ91 ]
_üåë –ú–µ—Ç–∫–∞ —Ç—å–º—ã_ *quisquam*"
`;
